plugin/client.go

```diff
@@ -39,3 +48,61 @@ type MattermostPlugin struct {
 func (p *MattermostPlugin) SetAPI(api API) {
 	p.API = api
 }
+
+// EnsureBot ether returns an existing bot user or creates a bot user with
+// the specifications of the passed bot.
+// Returns the id of the bot created or existing.
+func (p *MattermostPlugin) EnsureBot(bot *model.Bot) (retBotId string, retErr error) {
+	// Must provide a bot with a username
+	if bot == nil || len(bot.Username) < 1 {
+		return "", fmt.Errorf("EnsureBot was passed a bad bot")
+	}
+
+	// If we fail for any reason, this could be a race between creation of bot and
+	// retreval from anouther EnsureBot. Just try the basic retrieve existing again.
+	defer func() {
+		if retBotId == "" || retErr != nil {
+			time.Sleep(time.Second)
+			botIdBytes, err := p.API.KVGet(BOT_USER_KEY)
+			if err == nil {
+				retBotId = string(botIdBytes)
+				retErr = nil
+			}
+		}
+	}()
+
+	botIdBytes, err := p.API.KVGet(BOT_USER_KEY)
+	if err != nil {
+		return "", err
+	}
+
+	// If the bot has already been created, there is nothing to do.
+	if botIdBytes != nil {
+		botId := string(botIdBytes)
+		p.API.PatchBot(botId, &model.BotPatch{
+			Username:    &bot.Username,
+			DisplayName: &bot.DisplayName,
+			Description: &bot.Description,
+		})
+		return string(botIdBytes), nil
+	}
+
+	// Check for an existing bot user with that username. If one exists, then use that.
+	if user, err := p.API.GetUserByUsername(bot.Username); err == nil && user != nil {
+		if user.IsBot {
+			p.API.KVSet(BOT_USER_KEY, []byte(user.Id))
+			return user.Id, nil
+		}
+	}
+
+	// Create a new bot user for the plugin
+	createdBot, err := p.API.CreateBot(bot)
+	if err != nil {
+		p.API.LogError("Failed to create bot user.", "error", err)
```

Comment:
3/5 that the error should just be returned and not logged.
2/5 that you should to pass additional context via `errors.Wrap` instead.